src/runtime/recipe/handle.ts

```diff
@@ -180,8 +180,8 @@ export class Handle {
   get immediateValue() { return this._immediateValue; }
   set immediateValue(value: ParticleSpec) { this._immediateValue = value; }
 
-  static effectiveType(handleType: Type, connections: {type: Type, direction: string}[]) {
-    const variableMap = new Map();
+  static effectiveType(handleType: Type, connections: {type: Type|null|undefined, direction: string|undefined|null}[]) {
```

Comment:
why is there both here `null` and `undefined`?

Reply:
handle-connection has it defined that way.  The type there can be undefined (never set) or explicitly set to null.  Annoying..  Would like to root out usages like this..